File: python/group_anagrams.py
from typing import List
from collections import defaultdict

def group_anagrams(strs: List[str]) -> List[List[str]]:
    groups = defaultdict(list)

    for s in strs:
        # 排序后的字符串作为 key
        key = "".join(sorted(s))
        groups[key].append(s)

    return list(groups.values())


# 测试
print(group_anagrams(["eat","tea","tan","ate","nat","bat"]))
# [['eat', 'tea', 'ate'], ['tan', 'nat'], ['bat']]
print(group_anagrams([""]))      # [[""]]
print(group_anagrams(["a"]))     # [["a"]]

File: TS/group_anagrams.ts
function groupAnagrams(strs: string[]): string[][] {
    const groups = new Map<string, string[]>();

    for (const s of strs) {
        const key = s.split("").sort().join("");
        if (!groups.has(key)) {
            groups.set(key, []);
        }
        groups.get(key)!.push(s);
    }

    return Array.from(groups.values());
}

// 测试
console.log(groupAnagrams(["eat","tea","tan","ate","nat","bat"]));
// [['eat', 'tea', 'ate'], ['tan', 'nat'], ['bat']]
console.log(groupAnagrams([""]));      // [[""]]
console.log(groupAnagrams(["a"]));     // [["a"]]
